### PR TITLE
More open Event model

### PR DESCRIPTION
**Pull request by YYChen01988**
## Goal

Open up more Event model APIs, that users can add new Errors, Threads, and leave breadcrumbs at an `Event` level.

## Changeset

The `Event` class includes new methods to:
- add a new `Thread` object to the `Event`
- add a new `Error` object to the `Event`
- leave additional `Event` level breadcrumbs

The `Error` and `Thread` classes include new methods to add `Stackframe`s.

## Testing
New unit tests added.

### CODE REVIEW

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java [hunk 1]
```diff
@@ -98,4 +98,16 @@ static List<Error> createError(@NonNull Throwable exc,
                                    @NonNull Logger logger) {
         return ErrorInternal.Companion.createError(exc, projectPackages, logger);
     }
-}
+
+    @Nullable
+    public Stackframe addStackframe(@NonNull StackTraceElement element) {
+        return impl.addStackframe(element, logger);
+    }
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java:105]: This needs a null-check for the `element`.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java [hunk 2]
```diff
@@ -98,4 +98,16 @@ static List<Error> createError(@NonNull Throwable exc,
...
+
+    @Nullable
+    public Stackframe addStackframe(@NonNull String method,
+                                    @NonNull String file,
+                                    @NonNull Number lineNumber) {
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java:110]: These arguments should be `@Nullable` (as they are in the `Stackframe` constructor)

File: bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt [hunk 3]
```diff
@@ -32,4 +36,19 @@ internal class ErrorInternal @JvmOverloads internal constructor(
         writer.name("stacktrace").value(stacktrace)
         writer.endObject()
     }
+
+    fun addStackframe(element: StackTraceElement, logger: Logger): Stackframe? {
+        val frame = Stacktrace.serializeStackframe(element, emptyList(), logger)
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt:41]: This should use the `projectPackages` instead of `emptyList()`

File: bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt [hunk 4]
```diff
@@ -32,4 +36,19 @@ internal class ErrorInternal @JvmOverloads internal constructor(
...
+        if (frame != null) {
+            stacktrace.add(frame)
+            return frame
+        }
+        return null
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt:46]:
```suggestion
        frame?.let { stacktrace.add(it) }
        return frame
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt [hunk 5]
```diff
@@ -32,4 +36,19 @@ internal class ErrorInternal @JvmOverloads internal constructor(
...
+    }
+
+    fun addStackframe(method: String, file: String, lineNumber: Number): Stackframe {
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/ErrorInternal.kt:49]:
```suggestion
    fun addStackframe(method: String?, file: String?, lineNumber: Number?): Stackframe {
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java [hunk 6]
```diff
@@ -98,4 +98,16 @@ static List<Error> createError(@NonNull Throwable exc,
...
         return ErrorInternal.Companion.createError(exc, projectPackages, logger);
     }
-}
+
+    @Nullable
+    public Stackframe addStackframe(@NonNull StackTraceElement element) {
+        return impl.addStackframe(element, logger);
+    }
+
+    @Nullable
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Error.java:107]: The `Stackframe` returned by this function is not nullable

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 7]
```diff
@@ -52,7 +53,7 @@ private void logNull(String property) {
 
     /**
      * The Throwable object that caused the event in your application.
-     *
+     * <p>
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:56]: These formatting elements are new, and don't belong in this PR

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 8]
```diff
@@ -442,4 +443,46 @@ void setRedactedKeys(Collection<Pattern> redactedKeys) {
...
+    public Error addError(@NonNull String errorClass,
+                          @NonNull String errorMessage,
+                          @NonNull ErrorType errorType) {
+        return impl.addError(errorClass, errorMessage, errorType);
+    }
+
+    @Nullable
+    public Thread addThread(@NonNull String id,
+                            @NonNull String name) {
+        return impl.addThread(id, name, ErrorType.ANDROID, true, "RUNABLE");
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:474]:
```suggestion
        return impl.addThread(id, name, ErrorType.ANDROID, false, Thread.State.RUNNABLE.getDescriptor());
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 9]
```diff
@@ -442,4 +443,46 @@ void setRedactedKeys(Collection<Pattern> redactedKeys) {
...
+    }
+
+    @Nullable
+    public Breadcrumb leaveBreadcrumb(@NonNull String message,
+                                      @NonNull BreadcrumbType type,
+                                      @NonNull Map<String, Object> metadata) {
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:480]:
```suggestion
                                      @Nullable Map<String, Object> metadata) {
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 10]
```diff
@@ -442,4 +443,46 @@ void setRedactedKeys(Collection<Pattern> redactedKeys) {
...
+        return impl.addError(errorClass, errorMessage, errorType);
+    }
+
+    @Nullable
+    public Thread addThread(@NonNull String id,
+                            @NonNull String name) {
+        return impl.addThread(id, name, ErrorType.ANDROID, true, "RUNABLE");
+    }
+
+    @Nullable
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:477]: `leaveBreadcrumb` is not nullable

File: bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt [hunk 11]
```diff
@@ -324,4 +325,60 @@ internal class EventInternal : FeatureFlagAware, JsonStream.Streamable, Metadata
     override fun clearFeatureFlag(name: String) = featureFlags.clearFeatureFlag(name)
 
     override fun clearFeatureFlags() = featureFlags.clearFeatureFlags()
+
+    fun addError(throwError: Throwable): Error? {
+        val error = Error.createError(throwError, projectPackages, logger)
+        errors.addAll(error)
+        return error.firstOrNull()
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt:332]:
```suggestion
        val errors = Error.createError(throwError, projectPackages, logger)
        errors.addAll(error)
        return errors.firstOrNull()
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt [hunk 12]
```diff
@@ -324,4 +325,60 @@ internal class EventInternal : FeatureFlagAware, JsonStream.Streamable, Metadata
...
+    }
+
+    fun addError(errorClass: String, errorMessage: String, errorType: ErrorType): Error {
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt:335]: `errorMessage` should be nullable
```suggestion
    fun addError(errorClass: String, errorMessage: String?, errorType: ErrorType): Error {
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt [hunk 13]
```diff
@@ -324,4 +325,60 @@ internal class EventInternal : FeatureFlagAware, JsonStream.Streamable, Metadata
...
+            logger
+        )
+        threads.add(thread)
+        return thread
+    }
+
+    fun leaveBreadcrumb(
+        message: String,
+        type: BreadcrumbType,
+        metadata: MutableMap<String, Any>,
```
Comment (lemnik) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt:369]:
```suggestion
        metadata: MutableMap<String, Any?>?,
```

File: examples/sdk-app-example/app/src/main/java/com/example/bugsnag/android/BaseCrashyActivity.kt [hunk 14]
```diff
@@ -9,6 +9,7 @@ import androidx.appcompat.app.AppCompatActivity
 import com.bugsnag.android.BreadcrumbType
 import com.bugsnag.android.Bugsnag
 import com.bugsnag.android.Configuration
+import com.bugsnag.android.ErrorType
```
Comment (lemnik) [inline examples/sdk-app-example/app/src/main/java/com/example/bugsnag/android/BaseCrashyActivity.kt:12]: Unused import?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt (hunk 11 above)
Comment (YYChen01988) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt]: I changed it to newErrors to avoid the name duplication

File: bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt [hunk 15]
```diff
@@ -324,4 +325,72 @@ internal class EventInternal : FeatureFlagAware, JsonStream.Streamable, Metadata
     override fun clearFeatureFlag(name: String) = featureFlags.clearFeatureFlag(name)
 
     override fun clearFeatureFlags() = featureFlags.clearFeatureFlags()
+
+    fun addError(thrownError: Throwable?): Error {
+        if (thrownError == null) {
+            val newError = Error(
+                ErrorInternal("null", null, Stacktrace(ArrayList())),
```
Comment (kstenerud) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/EventInternal.kt:332]: This feels weird, adding a "null" error when a null throwable is passed. When would this happen other than as a programming error?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Stacktrace.kt [hunk 16]
```diff
@@ -49,34 +73,10 @@ internal class Stacktrace : JsonStream.Streamable {
...
-                methodName,
-                el.fileName ?: "Unknown",
-                el.lineNumber,
-                inProject(className, projectPackages)
-            )
-        } catch (lineEx: Exception) {
-            logger.w("Failed to serialize stacktrace", lineEx)
-            return null
+            frames.size >= STACKTRACE_TRIM_LENGTH -> frames.subList(0, STACKTRACE_TRIM_LENGTH).toMutableList()
+            else -> frames.toMutableList()
```
Comment (kstenerud) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Stacktrace.kt:79]: This could start to generate a lot of heap data... Is it possible to introspect `frames` to see if it's already mutable?

**Review by kstenerud — APPROVED** [review]

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 17]
```diff
@@ -75,6 +75,35 @@ public List<Error> getErrors() {
         return impl.getErrors();
     }
 
+    /**
+     * Add a new error to this report and return its Error data. The new Error will appear at the
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:79]:
```suggestion
     * Add a new error to this event and return its Error data. The new Error will appear at the
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 18]
```diff
@@ -75,6 +75,35 @@ public List<Error> getErrors() {
...
+     * end of the {@link #getErrors() errors list}.
+     */
+    @NonNull
+    public Error addError(@NonNull Throwable error) {
+        return impl.addError(error);
+    }
+
+    /**
+     * Add a new empty {@link ErrorType#ANDROID android} error to this report and return its Error
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:88]:
```suggestion
     * Add a new empty {@link ErrorType#ANDROID android} event to this report and return its Error
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 19]
```diff
@@ -75,6 +75,35 @@ public List<Error> getErrors() {
...
+     * data. The new Error will appear at the end of the {@link #getErrors() errors list}.
+     */
+    @NonNull
+    public Error addError(@NonNull String errorClass, @Nullable String errorMessage) {
+        return impl.addError(errorClass, errorMessage, ErrorType.ANDROID);
+    }
+
+    /**
+     * Add a new empty error to this report and return its Error data. The new Error will appear
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:97]:
```suggestion
     * Add a new empty error to this event and return its Error data. The new Error will appear
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 20]
```diff
@@ -84,6 +113,46 @@ public List<Thread> getThreads() {
         return impl.getThreads();
     }
 
+    /**
+     * Create, add and return a new empty {@link Thread} object to this event with a given id
+     * and name. This can be used to augment the report with thread data that would not be picked
+     * up as part of a normal report being generated (for example: native threads managed
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:119]:
```suggestion
     * and name. This can be used to augment the event with thread data that would not be picked
     * up as part of a normal event being generated (for example: native threads managed
```

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java [hunk 21]
```diff
@@ -84,6 +113,46 @@ public List<Thread> getThreads() {
...
+                ErrorType.ANDROID,
+                false,
+                Thread.State.RUNNABLE.getDescriptor()
+        );
+    }
+
+    /**
+     * Create, add and return a new empty {@link Thread} object to this event with a given id
+     * and name. This can be used to augment the report with thread data that would not be picked
+     * up as part of a normal report being generated (for example: native threads managed
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/Event.java:139]:
```suggestion
     * and name. This can be used to augment the event with thread data that would not be picked
     * up as part of a normal event being generated (for example: native threads managed
```

File: CHANGELOG.md [hunk 22]
```diff
@@ -6,6 +6,8 @@
 
 * Include additional Intent information for Activity.onCreate breadcrumbs (action, categories, type, flags, id, extra keys)
   [#2057](https://github.com/bugsnag/bugsnag-android/pull/2057)
+* New APIs allowing new `Error`s, `Thread`s, and `Stackframe`s to be added to error reports (`Event`s)
```
Comment (tomlongridge) [inline CHANGELOG.md:9]:
```suggestion
* New APIs allowing new `Error`s, `Thread`s, and `Stackframe`s to be added to `Event`s
```

**Review by tomlongridge — APPROVED** [review]
I think calling them events throughout is better than the word report as that matches our docs. But don't feel strongly if you have a better reason.

**Review by yousif-bugsnag — APPROVED** [review]
LGTM
